src/core/meshcop/commissioner.cpp

```diff
@@ -164,27 +164,40 @@ otError Commissioner::Start(otCommissionerStateCallback  aStateCallback,
     SetState(OT_COMMISSIONER_STATE_PETITION);
 
 exit:
+    if (error != OT_ERROR_NONE)
+    {
+        Get<Coap::CoapSecure>().Stop();
+    }
     return error;
 }
 
-otError Commissioner::Stop(void)
+otError Commissioner::Stop(bool aResign)
 {
     otError error = OT_ERROR_NONE;
 
     VerifyOrExit(mState != OT_COMMISSIONER_STATE_DISABLED, error = OT_ERROR_INVALID_STATE);
 
     Get<Coap::CoapSecure>().Stop();
 
-    Get<ThreadNetif>().RemoveUnicastAddress(mCommissionerAloc);
-    RemoveCoapResources();
-    ClearJoiners();
-    mTransmitAttempts = 0;
+    if (mState == OT_COMMISSIONER_STATE_ACTIVE)
+    {
+        Get<ThreadNetif>().RemoveUnicastAddress(mCommissionerAloc);
+        RemoveCoapResources();
+        ClearJoiners();
+    }
+    else if (mState == OT_COMMISSIONER_STATE_PETITION)
+    {
+        mTransmitAttempts = 0;
+    }
 
     mTimer.Stop();
 
     SetState(OT_COMMISSIONER_STATE_DISABLED);
 
-    SendKeepAlive();
+    if (aResign)
```

Comment:
the resign keep alive should only required when stopping an active commissioner, not necessary if it is still candidate.

Reply:
There are two cases when an active commissioner can be stopped: 
* When commissioner session is rejected in KeepAlive response, the commissioner need to stop, but no need to `SendKeepAlive (reject)`. 
* When commissioner session is stopped by API, the commissioner need to send `SendKeepAlive (reject)`.

So, the commissioner can not determine if it's necessary to `SendKeepAlive()` only by checking state. `aResign` argument is required at least for the active state.

Reply:
I mean this resign here -  if stopped by API, there should not be `SendKeepAlive(reject)` if it is still in petition. 

for the existing code here, may introduce a bool variable to record whether or not it was an  active commissioner, and here the condition should be (aResign && boolean variable)
make sense?

Reply:
Yes, we can use more conditions to avoid this `KeepAlive reject`. 
If commissioner is stopped by API in petition state, it will send `KeepAlive reject` with current `mSessionId` which I believe should be 0 or last active session Id. Either case, the `KeepAlive reject` has incorrect session ID.

Reply:
@librasungirl I think the commissioner state perfectly indicates whether it is active. No need for extra flags. right? We just need to change commissioner state to `DISABLED` later.

Reply:
In exiting implementation, `SendKeepAlive()` would has state `kAccept` or `kReject` according to the commissioner state.
https://github.com/openthread/openthread/blob/master/src/core/meshcop/commissioner.cpp#L760

`DISABLED` should be set before call `SendKeepAlive()`

Reply:
@librasungirl You are right about the implementation. But I don't think it make sense to have a commissioner send a keep alive message when it is in `DISABLED` state. Once a commissioner became `DISABLED`, it shouldn't be able  to send any messages. Agree?

So I would suggest changing signature `SendKeepAlive()` to  `SendKeepAlive(bool aAccept)`, and moving `SetState(...)` to the end of this function. Thoughts?

Reply:
> So I would suggest changing signature `SendKeepAlive()` to `SendKeepAlive(bool aAccept)`, and moving `SetState(...)` to the end of this function. 

@wgtdkp I think you are right. The current `SendKeepAlive` implementation is **implicitly** using current commissioner state to determine if it's a reject or accept. Could be a surprise for people read the code.

Reply:
I think it's better to make this PR a dedicated fix for handling commissioning messages. 
@wgtdkp would you create another PR for refactoring `SendKeepAlive` prototype? 
Thoughts? @librasungirl @wgtdkp @bukepo

Reply:
SGTM.

Reply:
> for the existing code here, may introduce a bool variable to record whether or not it was an active commissioner, and here the condition should be (aResign && boolean variable)

added variable `needResign` for this purpose.